File: lab3/java/AdjacencyGraph.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;
import java.util.LinkedList;
import java.util.Set;
import java.util.Map;
import java.util.HashMap;

/**
 * This is a class for a generic finite graph, with string vertices.
 *  - The edges are stored as an adjacency list as described in the course book and the lectures.
 *  - The graphs can be anything, such as a road map or a web link graph.
 *  - The graph can be read from a simple text file with one edge per line.
 */

public class AdjacencyGraph implements Graph<String> {
    private final Map<String, List<Edge<String>>> adjacencyList;
    private boolean weighted;

    AdjacencyGraph() {
        this.adjacencyList = new HashMap<>();
        this.weighted = false;
    }

    AdjacencyGraph(String graph) throws IOException {
        this();
        this.init(graph);
    }

    /**
     * Populates the graph with edges from a text file.
     * The file should contain one edge per line, each on the form
     * "from \\t to \\t weight" or "from \\t to"(where \\t == TAB).
     */
    @Override
    public void init(String graph) throws IOException {
        Files.lines(Paths.get(graph))
            .map(String::strip)
            .filter(line -> !line.startsWith("#"))
            .map(line -> line.split("\t"))
            .map(parts -> ( parts.length == 2
                          ? new Edge<>(parts[0], parts[1])
                          : new Edge<>(parts[0], parts[1], Double.parseDouble(parts[2]))
                          ))
            .forEach(this::addEdge);
    }

    /**
     * Adds a vertex to this graph.
     */
    public void addVertex(String v) {
        this.adjacencyList.putIfAbsent(v, new LinkedList<>());
    }

    /**
     * Adds a directed edge (and its source and target vertices) to this edge-weighted graph.
     * Note: This does not test if the edge is already in the graph!
     */
    public void addEdge(Edge<String> e) {
        this.addVertex(e.start);
        this.addVertex(e.end);
        this.adjacencyList.get(e.start).add(e);
        if (!this.weighted && e.weight != 1)
            this.weighted = true;
    }

    @Override
    public Set<String> vertices() {
        return Collections.unmodifiableSet(this.adjacencyList.keySet());
    }

    @Override
    public List<Edge<String>> outgoingEdges(String v) {
        return Collections.unmodifiableList(this.adjacencyList.get(v));
    }

    public boolean isWeighted() {
        return this.weighted;
    }

    @Override
    public String parseVertex(String s) {
        if (!this.adjacencyList.containsKey(s)) 
            throw new IllegalArgumentException("Unknown vertex: " + s);
        return s;
    }

    @Override
    public String toString() {
        return (
            (this.weighted ? "Weighted" : "Unweighted") +
            String.format(" adjacency graph of dimensions %d vertices and %d edges\n", this.numVertices(), this.numEdges()) +
            "\nRandom example points with outgoing edges:\n" + 
            this.exampleOutgoingEdges(8)
        );
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) throw new IllegalArgumentException();
        AdjacencyGraph graph = new AdjacencyGraph(args[0]);
        System.out.println(graph);
    }

}

File: lab3/python/adjacency_graph.py
import sys
from typing import Dict, List, FrozenSet, Optional

from edge import Edge
from graph import Graph


Vertex = str

class AdjacencyGraph(Graph[Vertex]):
    """
    This is a class for a generic finite graph, with string vertices.
     - The edges are stored as an adjacency list as described in the course book and the lectures.
     - The graphs can be anything, such as a road map or a web link graph.
     - The graph can be read from a simple text file with one edge per line.
    """
    adjacencyList: Dict[Vertex, List[Edge[Vertex]]]
    weighted: bool

    def __init__(self, graph: Optional[str] = None):
        self.adjacencyList = {}
        self.weighted = False
        if graph:
            self.init(graph)
        

    def init(self, graph: str):
        """
        Populates the graph with edges from a text file.
        The file should contain one edge per line, each on the form
        "from \\t to \\t weight" or "from \\t to" (where \\t == TAB).
        """
        if graph:
            with open(graph, encoding="utf-8") as IN:
                for line in IN:
                    line = line.strip()
                    if line and not line.startswith('#'):
                        start, end, *maybe_weight = line.split('\t')
                        if (not maybe_weight):
                            self.addEdge(Edge(start, end))
                        else:
                            weight = float(maybe_weight[0])
                            self.addEdge(Edge(start, end, weight))

    def addVertex(self, v: Vertex):
        """Adds a vertex to this graph."""
        self.adjacencyList.setdefault(v, [])

    def addEdge(self, e: Edge[Vertex]):
        """
        Adds a directed edge (and its source and target vertices) to this edge-weighted graph.
        Note: This does not test if the edge is already in the graph!
        """
        self.addVertex(e.start)
        self.addVertex(e.end)
        self.adjacencyList[e.start].append(e)
        if not self.weighted and e.weight != 1:
            self.weighted = True

    def vertices(self) -> FrozenSet[Vertex]:
        return frozenset(self.adjacencyList)

    def outgoingEdges(self, v: Vertex) -> List[Edge[Vertex]]:
        return self.adjacencyList.get(v, [])

    def isWeighted(self) -> bool:
        return self.weighted

    def parseVertex(self, s: str):
        if s not in self.adjacencyList:
            raise ValueError(f"Unknown vertex: {s}")
        return s

    def __str__(self) -> str:
        return (
            ("Weighted" if self.weighted else "Unweighted") +
            f" adjacency graph with {self.numVertices()} vertices and {self.numEdges()} edges.\n" +
            f"\nRandom vertices with outgoing edges:\n" +
            self.exampleOutgoingEdges(8)
        )


if __name__ == '__main__':
    _, file = sys.argv
    graph = AdjacencyGraph(file)
    print(graph)

